File: tools/verify/render_meta.js
// Reference side for the cover art and lyrics screens.
//
// These read META/POS/LYR/UI globals and helpers that live in app.js, so the
// harness cuts the helpers out by brace matching and supplies the globals.
// vizCover/vizLyrics themselves come from viz.js unmodified.
const fs = require("fs");
const path = require("path");
const vm = require("vm");

const W = 192, H = 48;
const ROOT = path.join(__dirname, "..", "..");
const px = new Uint8Array(W * H);
const appSrc = fs.readFileSync(path.join(ROOT, "legacy", "web", "app.js"), "utf8");

function extract(name) {
  const start = appSrc.indexOf(`function ${name}(`);
  if (start < 0) throw new Error(`${name} not found`);
  let depth = 0;
  for (let j = appSrc.indexOf("{", start); j < appSrc.length; j++) {
    if (appSrc[j] === "{") depth++;
    else if (appSrc[j] === "}") { depth--; if (depth === 0) return appSrc.slice(start, j + 1); }
  }
  throw new Error(`unbalanced ${name}`);
}

const ctx = {
  GRID_W: W, GRID_H: H, LYRIC_CELLS: 30,
  setDot(f, x, y, i) {
    if (x < 0 || x >= W || y < 0 || y >= H) return;
    const idx = y * W + x;
    if (i > f[idx]) f[idx] = i;
  },
  oceanFrame() {},
  META: {}, POS: {}, LYR: {}, UI: { coverScroll: { offset: 0, phase: 0, t: 0 } },
  // Position is supplied already interpolated, matching what the core is given.
  trackPos() { return ctx.POS.base; },
  clockStr() { return "00:00"; },   // never reached: cover cases all set a duration
};
ctx.globalThis = ctx;
vm.createContext(ctx);
vm.runInContext(fs.readFileSync(path.join(ROOT, "legacy", "web", "font.js"), "utf8"), ctx);
vm.runInContext([extract("deckText"), extract("wrapLyric"), extract("scrollText"),
                 extract("lyricIndexAt")].join("\n"), ctx);
vm.runInContext(fs.readFileSync(path.join(ROOT, "legacy", "web", "viz.js"), "utf8"), ctx);
vm.runInContext("globalThis.__cover = vizCover; globalThis.__lyrics = vizLyrics;", ctx);

function fnv1a(buf) {
  let h = 2166136261 >>> 0;
  for (const b of buf) { h = (h ^ b) >>> 0; h = Math.imul(h, 16777619) >>> 0; }
  return h >>> 0;
}

// Deterministic art and bands, generated the same way as the C harness.
let s = 0;
const next = () => { s = (Math.imul(s, 1664525) + 1013904223) >>> 0; return s / 4294967296; };

const cases = fs.readFileSync(
  process.argv[2] || path.join(ROOT, "tools", "verify", "meta.tsv"), "utf8");

for (const raw of cases.split("\n")) {
  if (!raw || raw.startsWith("/")) continue;
  const f = raw.replace(/\r$/, "").split("\t");
  if (f.length < 12) continue;
  const [name, screen, seed, posS, durS, stateS, syncedS, offS, title, artist, album, app] = f;
  const lyrics = f.slice(12).join("\t");

  s = Number(seed) >>> 0;
  const V = { bands: [], peaks: [], bassAvg: 0 };
  for (let b = 0; b < 13; b++) V.bands.push(next());
  for (let b = 0; b < 13; b++) V.peaks.push(next());
  V.bassAvg = next();

  // 48x48 art, dithered the same way ditherSquare does
  const S = 48, lum = new Uint8Array(S * S);
  for (let i = 0; i < S * S; i++) lum[i] = Math.floor(next() * 256);
  let lo = 255, hi = 0;
  for (const l of lum) { if (l < lo) lo = l; if (l > hi) hi = l; }
  const span = Math.max(30, hi - lo);
  const BAYER4 = [[0, 8, 2, 10], [12, 4, 14, 6], [3, 11, 1, 9], [15, 7, 13, 5]];
  const art = new Uint8Array(S * S);
  for (let y = 0; y < S; y++)
    for (let x = 0; x < S; x++) {
      const v = ((lum[y * S + x] - lo) / span) * 3.999;
      const t = (BAYER4[y & 3][x & 3] + 0.5) / 16;
      art[y * S + x] = Math.max(0, Math.min(3, Math.floor(v + t - 0.5)));
    }

  ctx.META = {
    title: ctx.deckText(title), artist: ctx.deckText(artist),
    album: ctx.deckText(album), app: ctx.deckText(app),
    artBig: Number(seed) === 0 ? null : art,
  };
  ctx.POS = { base: Number(posS), duration: Number(durS), status: "paused", at: 1 };
  ctx.UI.coverScroll = { offset: 0, phase: 0, t: 0 };

  const state = ["idle", "searching", "ok", "none"][Number(stateS)];
  const lines = lyrics ? lyrics.split(";").filter(Boolean).map((e) => {
    const i = e.indexOf("|");
    return [Number(e.slice(0, i)), ctx.deckText(e.slice(i + 1))];
  }) : [];
  const rows = [], rowStart = [];
  lines.forEach((ln, i) => {
    rowStart[i] = rows.length;
    for (const text of ctx.wrapLyric(ln[1], 30)) rows.push({ i, text });
  });
  ctx.LYR = {
    state, synced: Number(syncedS) === 1, lines, rows, rowStart,
    offset: Number(offS), key: "k",
  };

  px.fill(0);
  if (screen === "cover") ctx.__cover(px, V, 100, 0);
  else ctx.__lyrics(px, V, 100, 0);

  let nz = 0, sum = 0;
  for (const b of px) { if (b) nz++; sum += b; }
  console.log(`${name.padEnd(16)} hash=${fnv1a(px).toString(16).padStart(8, "0")} ` +
              `nz=${String(nz).padEnd(5)} sum=${sum}`);
}

File: tools/verify/render_meta.c
/* C side for the cover art and lyrics screens. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../../core/deck.h"
#include "../../core/screens.h"
#include "../../core/art.h"
#include "../../core/text.h"

#define W 192
#define H 48

static uint32_t lcg_s;
static double lcg_next(void) {
  lcg_s = lcg_s * 1664525u + 1013904223u;
  return (double)lcg_s / 4294967296.0;
}

static uint32_t fnv1a(const uint8_t *p, size_t n) {
  uint32_t h = 2166136261u;
  for (size_t i = 0; i < n; i++) { h ^= p[i]; h *= 16777619u; }
  return h;
}

static void copy_folded(char *dst, size_t cap, const char *src) {
  deck_fold(src, dst, cap);
}

int main(int argc, char **argv) {
  const char *path = argc > 1 ? argv[1] : "tools/verify/meta.tsv";
  FILE *f = fopen(path, "r");
  if (!f) { perror(path); return 1; }

  static uint8_t px[W * H];
  static uint8_t lum[48 * 48], art[48 * 48];
  const deck_geom_t geom = { W, H, DECK_LEVELS, 0 };
  deck_fb_t fb = { &geom, px };
  static deck_state_t v;
  static deck_meta_t m;

  char line[4096];
  while (fgets(line, sizeof line, f)) {
    if (line[0] == '/' || line[0] == '\n') continue;
    size_t len = strlen(line);
    while (len && (line[len - 1] == '\n' || line[len - 1] == '\r')) line[--len] = 0;

    /* 12 tab-separated fields then the lyrics blob. A line whose trailing
     * fields are all empty has no final tab, so the last field may simply run
     * to end of line with an empty remainder — JS's split() yields those, and
     * this has to agree with it or cases silently disappear. */
    char *fld[12], *p = line;
    int ok = 1;
    for (int i = 0; i < 12; i++) {
      char *t = strchr(p, '\t');
      if (t) { *t = 0; fld[i] = p; p = t + 1; }
      else if (i == 11) { fld[i] = p; p = line + len; }
      else { ok = 0; break; }
    }
    if (!ok) continue;
    const char *name = fld[0], *screen = fld[1];
    const unsigned seed = (unsigned)strtoul(fld[2], NULL, 10);
    const char *lyrics = p;

    memset(&v, 0, sizeof v);
    memset(&m, 0, sizeof m);
    lcg_s = seed;
    for (int b = 0; b < DECK_BANDS; b++) v.bands[b] = lcg_next();
    for (int b = 0; b < DECK_BANDS; b++) v.peaks[b] = lcg_next();
    v.bassAvg = lcg_next();

    const int S = 48;
    for (int i = 0; i < S * S; i++) lum[i] = (uint8_t)(lcg_next() * 256.0);
    deck_art_dither(lum, S, art);

    copy_folded(m.title,  sizeof m.title,  fld[8]);
    copy_folded(m.artist, sizeof m.artist, fld[9]);
    copy_folded(m.album,  sizeof m.album,  fld[10]);
    copy_folded(m.app,    sizeof m.app,    fld[11]);
    m.position   = atof(fld[3]);
    m.duration   = atof(fld[4]);
    m.lyricState = atoi(fld[5]);
    m.synced     = atoi(fld[6]);
    m.offsetMs   = atof(fld[7]);
    m.art     = seed == 0 ? NULL : art;
    m.artSide = seed == 0 ? 0 : S;

    /* lyrics blob: "time|text;time|text;..." */
    char blob[3072];
    strncpy(blob, lyrics, sizeof blob - 1);
    blob[sizeof blob - 1] = 0;
    char *save = blob;
    while (*save) {
      char *semi = strchr(save, ';');
      if (semi) *semi = 0;
      char *bar = strchr(save, '|');
      if (bar) {
        *bar = 0;
        deck_lyrics_add(&m, atof(save), bar + 1, 30);
      }
      if (!semi) break;
      save = semi + 1;
    }

    deck_clear(&fb);
    if (strcmp(screen, "cover") == 0) {
      deck_scroll_t sc = { 0, 0, 0.0 };
      deck_screen_cover(&fb, &v, &m, &sc, 100.0);
    } else {
      deck_screen_lyrics(&fb, &v, &m, 0.0);
    }

    int nz = 0; long sum = 0;
    for (int i = 0; i < W * H; i++) { if (px[i]) nz++; sum += px[i]; }
    printf("%-16s hash=%08x nz=%-5d sum=%ld\n", name, fnv1a(px, sizeof px), nz, sum);
  }
  fclose(f);
  return 0;
}
